### DISCUSSION TITLE
Subfolder custom sort not working

### DISCUSSION THREAD

**Discussion by Ricky-Kali** [SebastianMC/obsidian-custom-sort#69] [Q&A] [answered]
Hi, 

First of all, thank you for this wonderful plugin! 

This might be a stupid question, but I am struggling to make custom sort work for one my subfolder with space in name. My folder structure looks like this:

Vault:
/root

-  /test folder lol

    ---file1
    ---file2

- /folder1

    ---file3
    ---file4

I place the "sortspec" file in the same directory as root folder.

```
---
sorting-spec: |
  target-folder: root
  fold...
  test...
  target-folder: root/folder1
  file4
  file3
  target-folder: root/test folder lol
  file2
  file1
---
```

Somehow the sort works for files in "folder1" but not "test folder lol". I also tried "target-folder: root/test..." and "target-folder: root/test...lol" but none of them seems to solve the problem. 

Vault:
root

-   /folder1

    ---file4
    ---file3

-   /test folder lol

    ---file1
    ---file2

Please let me know if I am missing something. Thank you in advance.

**Comment by SebastianMC** [comment, marked as answer]
Hi @Ricky-Kali ,

I created a simple test vault from scratch to reproduce your example. \
It looks like below:

![60-initial](https://user-images.githubusercontent.com/23032356/224947371-afe07bed-c575-47d1-9ee5-8530bf13087f.jpg)

Then I copied & applied the exact sorting specification which you provided and it works as expected:

![60-sorted](https://user-images.githubusercontent.com/23032356/224947960-7b518a20-a171-459e-838d-a3378f12eafe.jpg)

...which means to determine why it is not working on your side some debugging could help.

Please modify your sorting spec to look exactly like below:

```
---
sorting-spec: |
  target-folder: regexp: debug: ^root$
  fold...
  test...
  target-folder: regexp: debug: root/folder1
  file4
  file3
  target-folder: regexp: debug: root/test folder lol
  file2
  file1
---
```

As you can see it is about transforming the plain-text matching into regexp-based specs for `target-folder:` with debug enabled to see how actually the specs match the underlying folders. \
After opening the Obsidian console and applying the updated sorting spec, the debug output looks like:

![60-debug-output](https://user-images.githubusercontent.com/23032356/224948961-60b1025a-6fa2-4fbc-9d14-b1af838ee70f.jpg)

Please do the same on your side and:
1. see if the regexp-based specs fixed the issue (it could)
2. please copy here the debug output generated in Obsidian console - it could allow for some more insight on what is happening in your vault

From my perspective, as the author of the plugin, the key thing is to determine why the sorting doesn't work on your side? And if the reason lies in the plugin code, fix it. That's why I appreciate your feedback, especially after the issue will be fixed

**Reply by Ricky-Kali** [reply]
Hi @SebastianMC,

Thank you for your response! Somehow Github didn't notify me so I just saw it today.

I really appreciate your effort to reproduce my example. Not sure why but today the sorting works like a charm, with or without the `regexp` option. The other day I tried everything from copying the folder name to using regexp  and still struggled to make it work. My theory is maybe the encoding formats of space make a difference but we can not tell on the screen? Just my theory.

Anyway, as much as I hope I can be of more help, the only issue I can identify now is the wildcard matching (`...`) sometimes doesn't function as expected. Below is an example.

There are 2 folders under root directory - `Privilege Escalation` and `test folder lol`.
![image](https://user-images.githubusercontent.com/31872198/226114860-986594c8-20e1-4c6d-87e4-ce10c1274e8f.png)

The problem is with `Privilege Escalation`. If I define the target folder in the spec file using complete name, sorting works.
```
  target-folder: regexp: debug: root/Privilege Escalation
  Windows
  target-folder: regexp: debug: root/tes...ol
  random
```
![image](https://user-images.githubusercontent.com/31872198/226115048-e5db7365-511e-404c-a3f7-6bd0d3069793.png)


Here's the console log.
![image](https://user-images.githubusercontent.com/31872198/226114992-424d27cc-d5e8-4b5b-9ab4-3c40704c6306.png)

However, if I go with the similar regexp pattern used for `test folder lol`, matching failed and no log shown.

```
  target-folder: regexp: debug: root/Privi...ion
  Windows
  target-folder: regexp: debug: root/tes...ol
  random
```
![image](https://user-images.githubusercontent.com/31872198/226115147-a213a32d-8e5a-4f72-9e8c-202d6e4f087f.png)

Here's the log for this time.

![image](https://user-images.githubusercontent.com/31872198/226115160-aac472f7-e8b0-45b1-a5be-75b20428e7d8.png)

This is a weird behavior and I have no other folder's name matching the pattern or even starting with letter P. 

Let me know if you need more details about this. 

Thank you again for your response and the amazing plugin!

**Reply by SebastianMC** [reply]
Hi @Ricky-Kali,

after reading your message I scratched my head about how to give you more insight and not overwhelm and not confuse you more :-)
Here you go:
- the plain syntax of `target-folder:` treats the rest of line literally as the full path of the folder. No regexps, no wildcards support
  - for example `target-folder: Some...folder` will match the folder named `Some...folder` (exactly so) in the root directory of the vault. And only that folder.
  - the characters `...` are interpreted literally as a component of the folder path
- adding the modifier `name:` changes the behavior slightly to match the folder name only. Still no regexps, no wildcards support
  - for example `target-folder: name: Some...folder` will match all folders named `Some...folder` (exactly so) in any location of the vault folders hierarchy
  - the characters `...` are interpreted literally as a component of the folder name
- adding the modifier `regexp:` switches the mechanics to work in a wholly different way and treats the rest of the line as JS-style regexp
  - for example `target-folder: regexp: Some...folder` will match all folders which contain in their path the text 'Some' followed by three characters (any characters, dot '.' symbol in JS regexp matches any charater) and then the word 'folder'\
  Examples of matching folder paths include:
    - `Some---folder` in any location in the vault
    - `Some X folder` in any location in the vault
    - `References/Some/A folder` (tricky! the path '/' character is matched by the '.' JS regex symbol as well)
    - `This is some B folder with data` (tricky! matching is case-insensitive, and a substring of folder name is matched)
- to match folders with name starting with 'Some' and ending with 'folder' the following regexp-based line has to be put in the sorting configuration:
  - `target-folder: regexp: ^Some.+data$
    - you can see the special regexp tokens: 
      - `^` at the beginning to indicate that the folder path must start with 'Some' (which follows the '^')
      - `.+` which matches any sequence of any characters, at least one character
      - `$` at the end to indicate, that the 'data' must be at the end of the folder path

Let me stop here with hope that I helped and not confused you more :-)
In your specific examples, regexp-based, you should replace `...` with `.+` and use `^` as the prefix and `$` as the suffix.\
For instance:
- `target-folder: regexp: debug: root/tes...ol` --> `target-folder: regexp: debug: ^root/tes.+ol$`

After confirming that it works as expected, you can also remove the `debug:` token to turn off console outputs

**Reply by Ricky-Kali** [reply]
Hi @SebastianMC ,

Ah! I see. So `target-folder: regexp: debug: root/tes...ol` matched folder `test folder lol` is an coincidence. It actually matched with the partial path 'test fol`. Thank you so much for the explanations. (I think) now I have a better understanding how different matching mechanisms work. You rock man!
